### DISCUSSION TITLE
Program Selection Modification

### DISCUSSION THREAD

**Discussion by Macabre2000** [pybricks/support#1387] [General]
I am fairly new to pybricks and I am trying to get code that allows the user to select from multiple programs.  I found the thread at the link [here ](https://github.com/pybricks/support/issues/861) #861 which provided a nice approach to allow for this. The only problem with the example code is that it exits after a selected code block is executed.  I tried modifying the code to allow the user to select a different code after the first selected code executes, but the next selected program does not seem to execute.  The code lets me select a new program letter, but does not execute when I press the circle button.  Are there any suggestions on how I could loop back to the program select menu after the previous selection completes?  

edit:: it appears that one issue i am having is that the motors are assigned within each sub-program which causes a conflict when the next program tries to connect to the port.  Is there a command i can use to delete the prior port associations so that the next program can assign them when it runs?

**Comment by Macabre2000** [comment]
```python
    from pybricks.hubs import PrimeHub
    from pybricks.parameters import Button, Color, Direction, Port, Stop
    from pybricks.tools import wait

    # Let's offer these menu options. You can add as many as you like.
    menu_options = ("D", "L", "F")
    menu_index = 0

    hub = PrimeHub()


    while True:

        # Normally, the center button stops the program. But we want to use the
        # center button for our menu. So we can disable the stop button.
        hub.system.set_stop_button(None)

        while True:

            hub.display.char(menu_options[menu_index])

            # Wait for any button.
            pressed = ()
            while not pressed:
                pressed = hub.buttons.pressed()
                wait(10)
        
            # Wait for the button to be released.
            while hub.buttons.pressed():
                wait(10)

            # Now check which button was pressed.
            if Button.CENTER in pressed:
                # Center button, so the menu is done!
                break
            elif Button.LEFT in pressed:
                # Left button, so decrement menu menu_index.
                menu_index = (menu_index + 1) % len(menu_options)
            elif Button.RIGHT in pressed:
                # Right button, so increment menu menu_index.
                menu_index = (menu_index - 1) % len(menu_options)

        # Now we want to use the Center button as the stop button again.
        hub.system.set_stop_button(Button.CENTER)

        # Based on the selection, choose a program.
        selected = menu_options[menu_index]
        if selected == "D":
            import Testtwo
        elif selected == "L":
            import fasdfasdf
        elif selected == "F":
            import rotationtest



```

**Reply by dlech** [reply]
In Python, imports only work once. If an `import` statement is called again, it doesn't run the script again because it is already imported.

So instead, you need to put everything inside a function instead of at the top level in your scripts, then call that function. You can pass things like the hub object and motor objects to the function so that you avoid the problem of trying to initialize them twice.

Untested, but something like this...

```python
# one.py

from pybricks.parameters import Color
from pybricks.tools import wait

def one(hub):
    hub.light.on(Color.GREEN)
    wait(5000)
```

```python
# two.py

from pybricks.parameters import Color
from pybricks.tools import wait

def two(hub):
    hub.light.on(Color.BLUE)
    wait(5000)
```

```python
# menu.py

from pybricks.hubs import PrimeHub
from pybricks.parameters import Color
from pybricks.tools import hub_menu

from one import one
from two import two

# initialize shared objects here
hub = PrimeHub()

while True:
    selected = hub_menu("1", "2", "X")
    
    try:
        if selected == "1":
            one(hub)
        elif selected == "2":
            two(hub)
        elif selected == "X":
            break
    # this will catch SystemExit caused by pressing the stop
    # button as well as any other unhandled exception
    except BaseException as ex:
        # if there are motors, stop them here
        # and do any other reset needed
        hub.light.on(Color.RED)
        # for debugging purposes
        print(ex)
```

**Reply by blushingbacon** [reply]
Is there a way to do this with code blocks?

**Reply by laurensvalk** [reply]
Yes, please see https://github.com/orgs/pybricks/discussions/1717. Thanks!

**Reply by WillWang99** [reply]
When we trid to define these two functions as async functions, the hub_menu cannot work as expected. Is there any demos to deal with the async functions for the menu selection?

**Comment by bcfletcher** [comment]
this is how I did the hub menu for our FLL team. seems to work well
```
# Team 2 Submerged Main menu
#
# menu revision number
menu_revision="15"
#
from pybricks.hubs import PrimeHub
from pybricks.pupdevices import Motor
from pybricks.parameters import Button, Color, Direction, Port, Side, Stop
from pybricks.robotics import DriveBase
from pybricks.tools import wait, StopWatch, multitask, run_task, hub_menu
from umath import sin, pi
from SM_robot_core_rev13 import stopEverything
from SM_t0_rev10 import T0_Run
from SM_t1_rev16 import T1_Run
from SM_t2_rev25 import T2_Run
from SM_t3_rev10 import T3_Run
from SM_t4_rev15 import T4_Run
from SM_t5_rev37 import T5_Run
from SM_t6_rev6 import T6_Run
from SM_t7_rev7 import T7_Run

print("Submerged Menu, rev#",menu_revision)
print("    Setting Up...")
# Initialize the hub.
hub = PrimeHub()

hub.system.set_stop_button(Button.CENTER)

# Make a menu to choose a letter. You can also use numbers.

# Based on the selection, run a program.
print("    Running Menu")
hub.light.on(Color.BLUE)

while True:
    selected = hub_menu("M","0","1","2","3","4","5","6","7","X")

    try:
        if selected == "M":
            break
        if selected == "0":
            wait(1000)
            run_task(T0_Run())
            run_task(stopEverything())
        elif selected == "1":
            wait(1000)
            run_task(T1_Run())
            run_task(stopEverything())
        elif selected == "2":
            wait(1000)
            run_task(T2_Run())  
            run_task(stopEverything())
        elif selected == "3":
            wait(1000)
            run_task(T3_Run())  
            run_task(stopEverything())
        elif selected == "4":
            wait(1000)
            run_task(T4_Run())  
            run_task(stopEverything())
        elif selected == "5":
            wait(1000)
            run_task(T5_Run())  
            run_task(stopEverything())
        elif selected == "6":
            wait(1000)
            run_task(T6_Run())  
            run_task(stopEverything())
        elif selected == "7":
            wait(1000)
            run_task(T7_Run())  
            run_task(stopEverything())            
        elif selected == "X":
            run_task(stopEverything())
            break

    except BaseException as menuException:
        print("Stop was Pressed or a Critical Error Occured. Stopping all motors.")
        print(menuException)
        run_task(stopEverything())
        break
    print("    Back to Menu")

print ("End of Menu - if you see this, there might be a problem.")
```


and the stopEverything() function is:

```
async def stopEverything():
    if debugL1: print("Stop All Motors")
    driveBase.use_gyro(False)
    leftMotor.stop()
    wait(250)
    rightMotor.stop()
    wait(250)
    topAttachment.stop()
    wait(250)
    frontAttachment.stop()
    wait(250)
    if debugL1: print("All motors should be stopped")
```

**Reply by rajeshsubramanianus** [reply]
Hi @bcfletcher, Can you please tell what is this "SM_t0_rev7"? Is this a separate .py file? I am new to PyBricks trying to build a menu for competition. Thanks.

**Reply by bcfletcher** [reply]
The SM_t#_rev# are Pybricks Python files for the different tasks for our FLL 2024/2025's Submerged competition. Each Task is one or more missions.  

T# is the task number. Our final set was 7 tasks 0 to 6
Rev# is the revision number for the task which is our simplified version control since git is too complicated for the 6-8 graders and our laptop’s wifi can’t connect at the school.

'M' and 'X' menu items really don't do much. they are mostly there to help the kids know beginning and end of the menu list and that the robot is in menu mode (as opposed the mode represented by the small square block).

So SM_t0_rev7 is really SM_t0_rev7.py, all the python files need to be loaded into the pybricks GUI and you download them to all to the robots by pressing the Play (arrow button) while in the menu tab in the Pybricks User Interface.

we also download Pybricks and run it from our two laptops as a Chrome Web App. Works phenomenally well. No wifi required.

We also have a Microsoft outlook.com account and we use Microsoft's OneDrive to sync all the task/missions files and all the innovation project files between the our various laptops automatically (when they  are connected to WiFi). Actually works suprisingly well. The kids laptops are Windows 11 and the Coaches are a mix of Windows and Macs  and we can sync files between them without too many problems.

I updated the code sample to the latest version we are using and added the code for the stopEverything() function. This is critical to make sure the motors are stopped and gyro is off. if not, you run the risk of problems if they lift up the robot and the gyro is enabled.

**Reply by BertLindeman** [reply]
> So SM_t0_rev7 is really SM_t0_rev7.pl,

Typo? 😄 

So SM_t0_rev7 is really `SM_t0_rev7.py`

Great setup, by the way.

Bert

**Reply by bcfletcher** [reply]
Thanks. But no it’s not a typo. “From” entries to not have a .py extension on the “From” line but they will have a py extension inside the pybricks app and on disk.

update: I fixed the .pl in the explanation above. Sorry about that. Thanks for telling me.

**Reply by BertLindeman** [reply]
The ".pl" for the filename on disk would have ".py" as extension. That was my typo idea.

Bert

**Reply by rajeshsubramanianus** [reply]
Thanks @bcfletcher for the detailed answer. This helps. Nice versioning strategy!

**Reply by bcfletcher** [reply]
Thanks!

**Reply by rajeshsubramanianus** [reply]
Based on inputs from this post and another post, was able to create a menu which auto increments every time (both in spite of successful and unsuccessful execution). This will be helpful in FLL competitions. Thanks to all!

```python

######################################################################################
##################################      menu.py     ########################################
######################################################################################

from pybricks.parameters import Direction, Button, Port, Stop, Color
from pybricks.pupdevices import Motor
from pybricks.robotics import DriveBase
from pybricks.tools import multitask, run_task, wait, hub_menu
from pybricks.hubs import PrimeHub
#Below are the Pyhton code (.py files) which are extracted from block code (.py files)
# one, two, etc. are functions that reside inside those files  
from code_1 import one
from code_2 import two
from code_3 import three
from code_4 import four
from code_5 import five
from code_6 import six

# The main program starts here.
# Set up all devices.
prime_hub = PrimeHub()
left = Motor(Port.A, Direction.COUNTERCLOCKWISE)
right = Motor(Port.E, Direction.CLOCKWISE)
drive_base = DriveBase(left, right, 56, 112)
front_motor = Motor(Port.B, Direction.CLOCKWISE)
back_motor = Motor(Port.F, Direction.CLOCKWISE)

# Assign center button in the hub as STOP button
#prime_hub.system.set_stop_button(Button.CENTER)

# Read the last run number (route) stored in memory
last_route = [v for v in  prime_hub.system.storage(0, read=1)]

# Based on last run number (route) create a new menu (with first value in the menu as last run number + 1)
# Each number represents a robot run. User has to choose one using the hub center button
if last_route == [1]:
    selected = hub_menu("2", "3", "4", "5", "6", "1")
elif last_route == [2]:
    selected = hub_menu("3", "4", "5", "6", "1", "2")
elif last_route == [3]:
    selected = hub_menu("4", "5", "6", "1", "2", "3")
elif last_route == [4]:
    selected = hub_menu("5", "6", "1", "2", "3", "4")
elif last_route == [5]:
    selected = hub_menu("6", "1", "2", "3", "4", "5")
elif last_route == [6]:
    selected = hub_menu("1", "2", "3", "4", "5", "6")

# Based on the selection, run a program.
# Pass robot components that are initialized in this program as input to the fucntions one, two, etc.
try:
    if selected == "1":
        run_task(one(prime_hub, drive_base, front_motor, back_motor))
    elif selected == "2":
        run_task(two(prime_hub, drive_base, front_motor, back_motor))
    elif selected == "3":
        run_task(three(prime_hub, drive_base, front_motor, back_motor))
    elif selected == "4":
        run_task(four(prime_hub, drive_base, front_motor, back_motor))
    elif selected == "5":
        run_task(five(prime_hub, drive_base, front_motor, back_motor))
    elif selected == "6":
        run_task(six(prime_hub, drive_base, front_motor, back_motor))
except BaseException as menuException:
    print(menuException)

# Write the last selected run number to memory
val = ord(selected) - ord("0")
write_value = [val] 
prime_hub.system.storage(0, write=bytes(write_value))
print ("program execution completed")
 
```

**Reply by fancooling** [reply]
Hi, I'm coaching an FLL team and started using Pybricks this year.

I'm trying to follow your example to build a menu. In the menu.py, the first few lines are like,
`
from run1 import main as run1_main
from run2 import main as run2_main
`

Unfortunately the program seems to run immediately at "from run1 import main", and never reached the point for me to select menus.

I believe it's because in run1.py, it has 
`run_task(main())`

However, this is generated by Pybricks block coding automatically. The kids I coach don't know how to write text Python program. 

Any suggestions?

**Reply by bcfletcher** [reply]
Currently using block coding and hub_menu() have issues including the issue you found. 

The sources (.zip file) and associated slide deck at https://github.com/bcfletcher/lego_projects/tree/main/FLL/2025_MD_FLL_Coaches_Conf  should help

Best to use block coding to prototype solutions, they copy the generated python into a new file and make mods manually so it works with hub_menu() correctly.

Boyd

**Comment by agni2012** [comment]
I know this is kind of late but you can do
`exec(open("a_file.py").read())`
to run a file _multiple_ times, unlike import.

**Reply by BertLindeman** [reply]
Hi Agni,

To run an imported program or function multiple times, you can make a file that contains one or more functions.

Do not get confused by my longer filenames.

A small example to be imported with two functions: 
1. turn motor forward
2. turn motor backward 


 I named it `issue_1387_my_function.py` 
```python

# filename issue_1387_my_functions.py

from pybricks.pupdevices import Motor
from pybricks.tools import wait

def motor_fwd(motor):
    turn_angle = 360
    print("turn motor", turn_angle, "degrees")
    motor.run_angle(100, turn_angle)
    wait(1000)

def motor_bwd(motor):
    turn_angle = -360
    print("turn motor", turn_angle, "degrees")
    motor.run_angle(100, turn_angle)
    wait(1000)

```

The main program:
```python

# program name issue_1387_main_example.py

from pybricks.pupdevices import Motor
from pybricks.parameters import Port
from pybricks.tools import wait

#from pybricks.hubs import PrimeHub
import issue_1387_my_functions

# hub = PrimeHub()
motor_a = Motor(Port.A)

# Run the code multiple times
issue_1387_my_functions.motor_fwd(motor_a)  # First run
wait(500)
issue_1387_my_functions.motor_bwd(motor_a)  # Second run backward
wait(500)
issue_1387_my_functions.motor_fwd(motor_a)  # Third run

# Can be used in loops too
for i in range(5):
    issue_1387_my_functions.motor_bwd(motor_a)
    wait(250)
    issue_1387_my_functions.motor_fwd(motor_a)
    wait(250)

```

This way you can call the functions numerous times.

Hope this helps.

Bert

**Comment by agni2012** [comment]
Thank you for your response. I already did this though, and forgot to
delete the old 'exec' message it after I realized that Pybricks python does
not support what I did.

**Reply by BertLindeman** [reply]
No problem at all.

Thank you for your feedback.

Bert
